### DISCUSSION TITLE
Visualizing custom extension of nav_msgs/OccupancyGrids

### DISCUSSION THREAD

**Discussion by siddancha** [foxglove/community#49] [Q&A]
Hi all,

**Overview**
I want to implement a custom visualization in Foxglove, and am hoping to get advice from people/devs about the easiest way to do this! This is my first shot at extending Foxglove. I'm not very familiar with the codebase, and honestly, it seems a bit daunting at this point. Any hand-holding would be greatly appreciated!! :blush: 

**My use case**
I have a custom ROS message type that is an extension of `nav_msgs/OccupancyGrids`: [[LINK]](https://gist.github.com/siddancha/757a76a1c96f59c038050126c100ebb7).

Each cell contains both an occupancy probability and a 2D velocity. I have a function in mind that maps (occupancy, 2D velocity) to an HSV value. I would like Foxglove to be able to render this colorful map, which would look something like this (ignore the colorwheel at the bottom-right):

<img src="https://user-images.githubusercontent.com/6638091/155665010-60f1ffee-734c-4c0e-85d0-78cf872ef99e.png" width="350">

What would be the best way to implement this? I'm guessing this won't be a panel extension; rather I should directly modify the main codebase?

Thanks,
Sid

**Comment by siddancha** [comment]
**Update:** I duplicated the code that renders `OccupancyGrids`, to render `DynamicOccupancyGrids`. I am able to visualize just the occupancy component of `DyamicOccupancyGrids` by leaving the REGL code un-touched. The visualization works, and I am able to render my custom messages in grayscale! Now I need to modify the REGL code to convert velocities to colors.

My changes are here: https://github.com/siddancha/studio/commit/a69d9e8e33c9a7c8f5c3bc8cfeff0984b10d44c4

**Reply by amacneil** [reply]
Nice work. We've had some discussions about supporting either colored occupancy grids, or simply letting you render a full Image (in any existing supported format) at a given point in the 3d panel. However, it would require some work so won't happen quickly.
